server/src/test/java/org/elasticsearch/transport/RemoteClusterServiceTests.java

```diff
@@ -1399,12 +1403,73 @@ public void testRemoteClusterServiceNotEnabledGetRemoteClusterConnection() {
                 IllegalArgumentException.class,
                 () -> service.getRemoteClusterService().getRemoteClusterConnection("test")
             );
-            assertThat(e.getMessage(), equalTo("this node does not have the remote_cluster_client role"));
+            assertThat(e.getMessage(), equalTo("node [node-1] does not have the [remote_cluster_client] role"));
+        }
+    }
+
+    public void testRemoteClusterServiceEnsureClientIsEnabled() throws IOException {
+        final var nodeNameSettings = Settings.builder().put(Node.NODE_NAME_SETTING.getKey(), "node-1").build();
+
+        // Shouldn't throw when the remote cluster client role is enabled.
+        final var settingsWithRemoteClusterClientRole = Settings.builder()
+            .put(nodeNameSettings)
+            .put(onlyRole(DiscoveryNodeRole.REMOTE_CLUSTER_CLIENT_ROLE))
+            .build();
+        try (RemoteClusterService service = new RemoteClusterService(settingsWithRemoteClusterClientRole, null)) {
+            service.ensureClientIsEnabled();
+        }
+
+        // Expect throws when missing the remote cluster client role.
+        final var settingsWithoutRemoteClusterClientRole = Settings.builder()
+            .put(nodeNameSettings)
+            .put(removeRoles(Set.of(DiscoveryNodeRole.REMOTE_CLUSTER_CLIENT_ROLE)))
+            .build();
+        try (RemoteClusterService service = new RemoteClusterService(settingsWithoutRemoteClusterClientRole, null)) {
+            final var exception = expectThrows(IllegalArgumentException.class, service::ensureClientIsEnabled);
+            assertThat(exception.getMessage(), equalTo("node [node-1] does not have the [remote_cluster_client] role"));
+        }
+
+        // Expect throws when missing both the remote cluster client role and search node role when stateless is enabled.
+        final var statelessEnabledSettingsOnNonSearchNode = Settings.builder()
+            .put(nodeNameSettings)
+            .put(onlyRole(DiscoveryNodeRole.INDEX_ROLE))
+            .put(DiscoveryNode.STATELESS_ENABLED_SETTING_NAME, true)
+            .build();
+        try (RemoteClusterService service = new RemoteClusterService(statelessEnabledSettingsOnNonSearchNode, null)) {
+            final var exception = expectThrows(IllegalArgumentException.class, service::ensureClientIsEnabled);
+            assertThat(
+                exception.getMessage(),
+                equalTo(
+                    "node [node-1] must have the [remote_cluster_client] role or the [search] role "
+                        + "in stateless environments to use linked project client features"
+                )
+            );
+        }
+
+        // Shouldn't throw when stateless is enabled on a search node or a node with remote cluster client role.
```

Comment:
Can we add a 3rd case where both search node and remote cluster client role are enabled since this is actually the current default in serverless?